### ISSUE TITLE
Mypy fails to follow fastavro imports (when configured with --folow-imports=error)

### ISSUE THREAD

**Issue by slemouzy** [fastavro/fastavro#661]
Hi, my team is willing to type check as much as possible our python projects and we noticed a strange behavior with fastavro. While checking the types, mypy fails to follow fastavro imports (https://mypy.readthedocs.io/en/stable/running_mypy.html#following-imports).

Here is a simple example that exhibits the same behavior with python 3.8.13 and 3.9.15 (on a Debian + pyenv)

```python
import fastavro.write

with open('out.avro', 'wb') as outfile:
    fastavro.write.writer(
        fo = outfile,
        schema = {
            "type": "record",
            "name": "Root",
            "fields": [
                {"name": "foo", "type": "int"}
            ]
        },
        records = [{"foo":42}],
    )
```

A mypy run with a strict follow import behavior `python -m mypy --follow-import=error test.py` ends up with:

```
test.py:1: error: Import of "fastavro.write" ignored  [misc]
test.py:1: note: (Using --follow-imports=error, module not passed on command line)
test.py:1: error: Import of "fastavro" ignored  [misc]
Found 2 errors in 1 file (checked 1 source file)
```

Running mypy with a less strict behavior `python -m mypy --follow-import=normal test.py` does not raise anything.

```
Success: no issues found in 1 source file
```

As the type stubs have been added a long time ago (https://github.com/fastavro/fastavro/pull/438) , I suppose that some of them are now outdated or maybe as the default less strict mypy behavior does not raise anything, this issue has never been detected until now?

**Comment by scottbelden** [comment]
You are right that there are probably issues with the typing that aren't checked and caught very well. In fact there is [another issue](https://github.com/fastavro/fastavro/issues/477) that also talks about the typing not being well tested.

I'm a fan of typing and mypy, but I imagine the way this codebase is set up with conditional imports might make it difficult for mypy:

https://github.com/fastavro/fastavro/blob/9349de1f478fa4797f86555ef616be30dfeee29f/fastavro/write.py#L1-L4

Looking at the error above, I'm having trouble figuring out what would need to be changed to correct it.

**Comment by slemouzy** [comment]
Mhhh, those tricky imports might explain the failure, right. And I agree mypy is not so clear about its own follow import issue.

I'll try to find out some time to dig into this issue and will keep you in touch with my findings. Thanks @scottbelden :smiley:
